### PR TITLE
Refine #2900: export with XAI

### PR DESCRIPTION
**Pull request by sovrasov**
### Summary

This PR extends https://github.com/openvinotoolkit/training_extensions/pull/2900
The main rationale to these changes is that we need to have a single mechanism for introducing XAI into the model.
Custom forward only for export adds the second xai forward mechanism which is hard to test.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/entity/classification.py [hunk 1]
```diff
@@ -75,12 +80,61 @@ def head_forward_fn(self, x: torch.Tensor) -> torch.Tensor:
...
+        if self.explain_hook is not None:
+
+            class ModelWithExplainOutput(nn.Module):
+                def __init__(self, model: nn.Module, hook: BaseRecordingForwardHook) -> None:
+                    super().__init__()
+                    self.model = model
+                    self.hook = hook
+                    self.hook.reset()
+
+                def forward(self, x: torch.Tensor) -> tuple:
```
Comment (kprokofi) [inline src/otx/core/model/entity/classification.py:108]: Why can't we move this as global forward function for the entire ExplainableOTXClsModel?
We can call super().forward() and then check whether we have explain hook or not to return just logits or tuple consist of logits and saliency maps.
I just don't like much this wrapper class which substitute our OTXModel.

File: src/otx/core/model/entity/classification.py [hunk 2]
```diff
@@ -75,12 +80,61 @@ def head_forward_fn(self, x: torch.Tensor) -> torch.Tensor:
...
+                    self.hook.reset()
+                    return logits, saliency_maps
+
+            return ModelWithExplainOutput(super()._exportable_model, self.explain_hook)
+
+        return super()._exportable_model
+
+    @staticmethod
+    def _update_onnx_output_names(onnx_export_configuration: dict) -> None:
+        if "output_names" not in onnx_export_configuration:
```
Comment (kprokofi) [inline src/otx/core/model/entity/classification.py]: Where this method will be used? What if we don't want to export saliency maps. Will this OK to append both ["logits", "saliency_map"] for output_names?

File: src/otx/engine/engine.py [hunk 3]
```diff
@@ -367,13 +367,15 @@ def export(
         checkpoint: str | Path | None = None,
         export_format: OTXExportFormatType = OTXExportFormatType.OPENVINO,
         export_precision: OTXPrecisionType = OTXPrecisionType.FP32,
+        dump_auxiliaries: bool = False,
```
Comment (kprokofi) [inline src/otx/engine/engine.py:370]: I don't really like the naming. Maybe "dump_features" or "dump_saliency_maps", or just "explain"

File: src/otx/core/model/entity/classification.py (hunk 1 above)
Comment (sovrasov) [reply src/otx/core/model/entity/classification.py:108]: The final mechanism of obtaining the saliency maps is still not clear. Actually, we can already access the hook in `_customize_outputs`, the problem is what to do with the raw saliency maps. This wrapper doesn't look as a comprehensive replacement to forward, it's rather a workaround if batch is 1.

File: src/otx/core/model/entity/classification.py (hunk 2 above)
Comment (sovrasov) [reply src/otx/core/model/entity/classification.py]: That's a leftover, deleted.

File: src/otx/engine/engine.py (hunk 3 above)
Comment (sovrasov) [reply src/otx/engine/engine.py:370]: This comes from the original implementation, I also prefer *explain, since feature vector is something unclear and required only because we have Geti integration requirements.
@negvet what do you think?

File: src/otx/core/model/entity/classification.py (hunk 1 above)
Comment (sovrasov) [reply src/otx/core/model/entity/classification.py:108]: Also, if we look in general, the whole idea of hooks is that we don't change anything in the forward itself, while can extend it with an arbitrary number of hooks. 
If we want a custom forward, we don't need hooks, but with hooks we have to generate a custom forward for export, since torch tracer is quite limited in terms of functionality.

Comment (vinnamkim) [reply src/otx/core/model/entity/classification.py:108]: https://github.com/openvinotoolkit/training_extensions/pull/2870#discussion_r1475416405
https://github.com/openvinotoolkit/training_extensions/pull/2900#discussion_r1487040737
Like these comments, I think that we should introduce a dedicated `forward()` function for XAI at this time.

Comment (sovrasov) [reply src/otx/core/model/entity/classification.py:108]: @vinnamkim does that mean we have to remove the hook? If we have a standalone forward, keeping the hooks makes no sense, like I described in the previous comment.

Comment (vinnamkim) [reply src/otx/core/model/entity/classification.py:108]: > @vinnamkim does that mean we have to remove the hook? If we have a standalone forward, keeping the hooks makes no sense, like I described in the previous comment.

Yes

I guess you guys can move forward gradually. For example,

1. Introduce `forward_explain()` with the hook

```python
def forward_explain(self, inputs) -> BatchPredictionWithXAI:
     handle = self.register_explain_hook()
     outputs = self.forward(inputs)
     augmented_outputs = augment_with_xai_outputs(outputs, handle.xai_outputs)
     handle.remove()
     return augmented_outputs
```
2. Remove using the hook from `forward_explain()`
